docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
+#. spin up a new Docker container based on the Ubuntu OS specified in the config file
+#. install all language dependencies from the cache
+#. create the environment and install docs' dependencies inside the container
+#. execute build commands inside the container
+#. push the output generated by build commands to the storage
 
-   Initial Dockerfile.* as example for this are pushed in this PR
-   https://github.com/readthedocs/readthedocs-docker-images/pull/166
+The main difference with the current approach are:
 
+* the image spin is selected depending on the OS version
+* all language dependencies are installed at build time
+* languages not offereing binaries are pre-compiled by Read the Docs and stored in the cache
+* miniconda/mambaforge are now managed with the same management tool
 
-Specifying extra user's dependencies
-------------------------------------
 
-Different users may have different requirements. We were already requested to install
-``swig``, ``imagemagick``, ``libmysqlclient-dev``, ``lmod``, ``rust``, ``poppler-utils``, etc.
+Specifying extra languages requirements
+---------------------------------------
 
-People with specific dependencies will be able to install them as APT packages or as extras
-using ``.readthedocs.yaml`` config file. Example:
+Different users may have different requirements.
+People with specific language dependencies will be able to install them by using ``.readthedocs.yaml`` config file.
+Example:
 
 .. code:: yaml
 
+   version: 3
    build:
-     image: ubuntu20
-     python: 3.9
-     system_packages:
-       - swig
-       - imagemagick
-     extras:
-       - node==14
-       - rust==1.46
+     os: ubuntu20
+     languages:
+       python: "3.9"  # supports "pypy3", "miniconda3" and "mambaforge"
+       nodejs: "14"
+       rust: "1.54.0"
+       golang: "1.17"
 
 Important highlights:
 
-* users won't be able to use custom Ubuntu PPAs to install packages
-* all APT packages installed will be from official Ubuntu repositories
-* not specifying ``build.image`` will pick the latest OS image available
-* not specifying ``build.python`` will pick the latest Python version available
-* Ubuntu 18 will still be available via ``stable`` and ``latest`` images
-* all ``node`` (major) pre-compiled versions on ``nodenv`` are available to select
-* all ``rust`` (minor) pre-compiled versions on ``rustup`` are available to select
-* knowing exactly what packages users are installing,
-  could allow us to prebuild extra images: ``ubuntu20-py37+node14``
+* do not treat Python language different from the others (will help us to support other non-Python doctools in the future)
+* specifying ``build.languages.python: "3"`` will use Python version ``3.x.y``, and may differ between builds
+* specifying ``build.languages.python: "3.9"`` will use Python version ``3.9.y``, and may differ between builds
+* specifying ``build.languages.nodejs: "14"`` will use nodejs version ``14.x.y``, and may differ between builds
```

Comment:
How are we updating what is installed when a user says `3.9`? Is it always the latest, or a specific version we maintain and manually switch (and then worry about breaking).

Reply:
I thought this to be a specific version we maintain and have control over. We can decide _when_ to upgrade (if ever required) and we can also roll back if the new pre-compiled version was built with a problem.

Installing always latest is harder to maintain because we have to build the new version each time a new patch version is released. So, we will be behind official releases. Besides, it will give projects different versions more often.

I don't really want to expose to the user the patch version, because that way we will have to cache many different versions ourselves and if the user selects one patched version that we don't have cached by mistake, those builds will add extra build time.

Reply:
We should note this in the doc 👍

Reply:
This was already covered in "How do we upgrade a Python version?" section. I added some more context there.